src/main/java/edu/kit/compiler/semantic/DetailedNameTypeAstVisitor.java

```diff
@@ -0,0 +1,505 @@
+package edu.kit.compiler.semantic;
+
+import java.util.Optional;
+
+import edu.kit.compiler.data.AstObject;
+import edu.kit.compiler.data.AstVisitor;
+import edu.kit.compiler.data.DataType;
+import edu.kit.compiler.data.DataType.DataTypeClass;
+import edu.kit.compiler.data.Operator.*;
+import edu.kit.compiler.data.ast_nodes.*;
+import edu.kit.compiler.data.ast_nodes.ClassNode.*;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.*;
+import edu.kit.compiler.data.ast_nodes.MethodNode.*;
+import edu.kit.compiler.data.ast_nodes.StatementNode.*;
+import edu.kit.compiler.lexer.StringTable;
+import edu.kit.compiler.semantic.NamespaceMapper.ClassNamespace;
+
+/**
+ * This AST visitor performs name analysis and type checking on method bodies.
+ * 
+ * Preconditions:
+ * - class namespaces can be found by name (= identifier)
+ * - each class namespace contains all fields and methods with name and type
+ * - builtin definitions (classes, fields, methods) are available
+ * 
+ * Postconditions:
+ * - all used variables reference their declaration
+ * - every field, method, statement and expression is correctly typed
+ * - no variable is declared twice inside the same method
+ * - every expression node contains a valid result type
+ * - static methods contain no reference to this
+ * - void is not used as a field, parameter or local variable type, in a new
+ * object or new array expression
+ * - hasError is set in all nodes where an error occured
+ * - user defined data types of fields and methods (incl. parameters) are valid
+ * - integer literal values are valid integer values (32-bit signed)
+ * 
+ * Not checked:
+ * - all code paths return a value
+ * - left side of an assignment is an l-value
+ */
```

Comment:
Is the check for whether `args` is used in a main method responsibility of the third pass?

Reply:
Also my job